pkg/pluginmanager/manager.go

```diff
@@ -402,6 +429,98 @@ func combineDuplicatePlugins(availablePlugins []discovery.Discovered) []discover
 	return selectedPlugins
 }
 
+func mergePluginEntries(plugin1, plugin2 *discovery.Discovered) *discovery.Discovered {
+	// Combine the installation status and installedVersion result when combining plugins
+	if plugin2.Status == common.PluginStatusInstalled {
+		plugin1.Status = common.PluginStatusInstalled
+	}
+	if plugin2.InstalledVersion != "" {
+		plugin1.InstalledVersion = plugin2.InstalledVersion
+	}
+
+	// Build a combined Source string
+	if plugin1.Source != plugin2.Source {
+		plugin1.Source = fmt.Sprintf("%s/%s", plugin1.Source, plugin2.Source)
+	}
+
+	// The discovery type could be OCI or Local.
+	// When dealing with different discovery types, unset it
+	if plugin1.DiscoveryType != plugin2.DiscoveryType {
+		plugin1.DiscoveryType = ""
+	}
+
+	artifacts1, ok := plugin1.Distribution.(distribution.Artifacts)
+	if !ok {
+		// This should not happened
+		log.Warningf("Plugin '%s' has an unexpected distribution type", plugin1.Name)
+		return plugin1
+	}
+
+	artifacts2, ok := plugin2.Distribution.(distribution.Artifacts)
+	if !ok {
+		// This should not happened
+		log.Warningf("Plugin '%s' has an unexpected distribution type", plugin2.Name)
+		return plugin1
+	}
+
+	// For every version in the second plugin, if it doesn't already exist
+	// in the first plugin, add it.
+	// Also build the new list of supported versions
+	for version := range artifacts2 {
+		_, exists := artifacts1[version]
+		if !exists {
+			artifacts1[version] = artifacts2[version]
+			plugin1.SupportedVersions = append(plugin1.SupportedVersions, version)
+		}
+	}
+	plugin1.Distribution = artifacts1
+	_ = utils.SortVersions(plugin1.SupportedVersions)
+
+	// Keep the following fields from the first plugin found
+	// - RecommendedVersion
+	// - Optional
+	// - ContextName
+	// - Scope
+	return plugin1
+}
+
+// mergeDuplicatePlugins combines the same plugins to eliminate duplicates by merging the information
+// of multiple entries of the same plugin into a single entry.  For example, the Central Repository can
+// provide details about a plugin, but an additional test discovery can provide other versions of the
+// same plugin.  This function will join all the information into one.
+// A plugin is determined by its name-target combination.
+// Note that if two versions of the same plugin are found more than once, it will be the first one
+// found that will be kept.  The order of the array "plugins" therefore matters.
+// This merge operation is deterministic due to the sequence of sources/plugins that we process always
+// being the same.
+func mergeDuplicatePlugins(plugins []discovery.Discovered) []discovery.Discovered {
```

Comment:
This new function is only called with the Central Repo feature enabled.  It may be interesting for the legacy case but I didn't want to spend time testing that case.